arrow-libs/fx/arrow-fx-coroutines/src/commonMain/kotlin/arrow/fx/coroutines/Resource.kt

```diff
@@ -16,84 +16,126 @@ import kotlin.coroutines.CoroutineContext
+ * }
+ * ```
+ *
+ * Here `releaseCase` also signals with what [ExitCase] state the `use` step finished.
+ *
+ * # Using and composing Resource
  *
  * ```kotlin:ank:playground
  * import arrow.fx.coroutines.*
  *
- * object Consumer
- * object Handle
- *
- * class Service(val handle: Handle, val consumer: Consumer)
+ * class UserProcessor {
+ *   fun start(): Unit = println("Creating UserProcessor")
+ *   fun shutdown(): Unit = println("Shutting down UserProcessor")
+ *   fun process(ds: DataSource): List<String> =
+ *    ds.users().map { "Processed $it" }
+ * }
  *
- * suspend fun createConsumer(): Consumer = Consumer.also { println("Creating consumer") }
- * suspend fun createDBHandle(): Handle = Handle.also { println("Creating db handle") }
- * suspend fun createFancyService(consumer: Consumer, handle: Handle): Service =
- *   Service(handle, consumer).also { println("Creating service") }
+ * class DataSource {
+ *   fun connect(): Unit = println("Connecting dataSource")
+ *   fun users(): List<String> = listOf("User-1", "User-2", "User-3")
+ *   fun close(): Unit = println("Closed dataSource")
+ * }
  *
- * suspend fun closeConsumer(consumer: Consumer): Unit = println("Closed consumer")
- * suspend fun closeDBHandle(handle: Handle): Unit = println("Closed db handle")
- * suspend fun shutDownFancyService(service: Service): Unit = println("Closed service")
+ * class Service(val db: DataSource, val userProcessor: UserProcessor) {
+ *   suspend fun processData(): List<String> = userProcessor.process(db)
+ * }
  *
  * //sampleStart
- * val resourceProgram = suspend {
- *   Resource(::createConsumer, ::closeConsumer)
- *     .zip(Resource(::createDBHandle, ::closeDBHandle))
- *     .flatMap { (consumer, handle) ->
- *       Resource({ createFancyService(consumer, handle) }, { service -> shutDownFancyService(service) })
- *     }.use { service ->
- *       // use service
- *       // <...>
- *       Unit
- *     }
+ * val userProcessor = resource {
+ *   UserProcessor().also(UserProcessor::start)
+ * } release UserProcessor::shutdown
+ *
+ * val dataSource = resource {
+ *   DataSource().also { it.connect() }
+ * } release DataSource::close
+ *
+ * suspend fun main(): Unit {
+ *   userProcessor.parZip(dataSource) { userProcessor, ds ->
+ *       Service(ds, userProcessor)
+ *     }.use { service -> service.processData() }
  * }
  * //sampleEnd
- *
- * suspend fun main(): Unit = resourceProgram.invoke()
  * ```
+ *
+ * [Resource]s are immutable that can be composed using [zip] or [parZip].
+ * [Resource]s guarantee that their release finalizers are always invoked in the correct order when an exception is raised or the context where the program is running gets canceled.
+ *
+ * To achieve this [Resource] ensures that the `acquire` & `release` step are [NonCancellable].
+ * If a cancellation signal, or an exception is received during `acquire` it immediately goes through since there is nothing yet to release.
```

Comment:
Imo this does not quite convey that a failure in `acquire` is taken as "the resource has not been allocated and thus does not need to be released". Not sure if that has changed since I last looked at it though :see_no_evil: 

```suggestion
 * If a cancellation signal, or an exception is received during `acquire`, the resource is assumed to not have been acquired and thus will not trigger the release function.
```

Reply:
Thanks for the super awesome addition!